### ISSUE TITLE
Stand up SkyPilot against the on-prem AMD GPU host

### ISSUE THREAD

**Issue by Zorro909** [Zorro909/skywright#21]
Part of #1

## Question

Nothing to decide here — this is manual work blocking a set of decisions. Until SkyPilot actually addresses the on-prem AMD host, `T1`, `T4` and `K5` are argued in the abstract.

Carry out the setup identified by the local-target research, then record as the resolution:

- What was done, and what had to be installed on the host.
- What SkyPilot reports about the GPUs, under what accelerator name, and at what requestable count.
- What a resource specification targeting it actually looks like.
- What did not work, and what remains fragile.
- Any credentials or configuration files created, and where they live.

Later tickets depend on these facts.

## Target host

The target is a **dedicated on-prem machine, not the developer workstation.** Nothing established about the workstation transfers to it — its OS, reachability, ROCm state and container runtime are all still to be established on the box itself.

- **2 × AMD Radeon RX 7800 XT**, 16 GB VRAM each. Navi 32, expected `gfx1101` — confirm with `rocm-smi` on the host rather than assuming.
- ROCm's support tier for consumer Navi 32 is looser than for the 7900 XTX, and consumer RDNA cards commonly need `HSA_OVERRIDE_GFX_VERSION` to satisfy PyTorch. **Establish a working `torch.cuda.is_available()` on the bare host first** — every Kubernetes and SkyPilot step below is wasted if the card cannot run ROCm PyTorch at all.

## What the route already implies

From [Is a local AMD/ROCm box a first-class SkyPilot target?](https://github.com/Zorro909/skywright/issues/2), the route is an on-prem Kubernetes context publishing `amd.com/gpu`, plus a hand-set `skypilot.co/accelerator` node label. Two consequences to carry in:

- **The AMD GPU Operator may be the wrong component.** SkyPilot's [AMD guide](https://docs.skypilot.ai/en/latest/reference/kubernetes/amd-gpu.html) prescribes cert-manager plus `rocm/gpu-operator-charts`, but that operator exists to manage *driver lifecycle* for Instinct fleets. If ROCm is already installed on this host, the plain [`ROCm/k8s-device-plugin`](https://github.com/ROCm/k8s-device-plugin) DaemonSet is the lighter match. **Unverified:** its README does not state which GPU models it enumerates — that consumer RDNA 3 is advertised at all is the single largest risk in this setup, and this ticket is where it gets tested.
- **Capability-shaped requests will not match this host.** SkyPilot expands `accelerators: 16GB+` through its static catalog, whose only AMD entries are `MI300X`, `Radeon MI25` and `Radeon Pro V520`. A hand-set label such as `rx7800xt` matches none of them, so only the literal name will schedule. Confirm this on the box — it is a fact `T3` needs whichever way it falls.

## Out of this ticket's remit

Two GPUs on one node is new to the map. Record what `amd.com/gpu` allocatable count the cluster publishes and whether `<name>:2` schedules — but do **not** settle here how many GPUs a Run may claim, or what the training contract says about the second card. That is a decision, and this ticket supplies facts only.

**Comment by Zorro909** [comment]
Context from [Spike: reach SkyPilot from a JVM process](https://github.com/Zorro909/skywright/issues/18), which stood SkyPilot up on this host incidentally. Recording it so this ticket doesn't re-discover it — none of this resolves #21, which still owns the AMD/ROCm target.

**Host as found (2026-08-08):** ROCm already live, `rocm-smi` reports two devices — `0x744c` (Navi 31, the discrete card) and `0x164e` (Raphael iGPU). Java 25 (GraalVM), `kubectl`, `mvn`/`gradle`, `uv` present. **No `sky`, no kind, and no Kubernetes contexts** before the spike.

**What was installed, all inside a session scratchpad, nothing system-wide:** `skypilot[kubernetes]` 0.13.0 into a throwaway `uv` venv on Python 3.12, and the `kind` v0.31.0 binary. Both removed with the cluster; the host is back as found.

**`docker` on this box is podman shimming as docker** (rootless, `podman info .Host.Security.Rootless == true`). `sky local up` nonetheless succeeded and created the kind cluster with 32 CPUs, and `sky check kubernetes` enabled the `kind-skypilot` context.

**The blocker #21 will hit: pods have no outbound TCP.** From inside a pod, `nc -z 1.1.1.1 443` and `nc -z <pypi-ip> 443` both fail; from the kind node the same targets connect fine. NAT itself is present and working (iptables-**nft**, masquerade counters incrementing — note `iptables` *legacy* in the node reports `table nat does not exist`, which is a red herring). DNS from pods resolves A records correctly, so this is egress, not resolution.

Consequence: **no SkyPilot job can provision on a kind cluster on this host**, because SkyPilot's runtime setup runs `uv pip install` inside the pod and cannot reach PyPI. The failure surfaces as `Failed to set up SkyPilot runtime on cluster` / `Pod … terminated: Error (exit code 1)`.

That will need resolving — or a non-kind path — before this ticket can report what SkyPilot sees of the AMD GPU. Worth noting `sky check` already warned it *"could not detect GPU/TPU resources (`amd.com/gpu`, …)"* on the kind cluster, which is expected: kind nodes are containers without the AMD GPU Operator that [Is a local AMD/ROCm box a first-class SkyPilot target?](https://github.com/Zorro909/skywright/issues/2) settled on.

**Comment by Zorro909** [comment]
Resolved on the dedicated host (2026-08-11).

## Host prerequisite

- Ubuntu 24.04.4 LTS, kernel `6.8.0-137-generic`.
- Docker Engine 29.7.1 was already installed and is a real rootful Docker daemon (the earlier Podman shim is gone).
- ROCm 7.2.0 was already installed.
- `rocm-smi` and `rocminfo` both report **2 × AMD Radeon RX 7800 XT**, `gfx1101`, with `17,163,091,968` bytes VRAM each.
- Native PyTorch was checked first in an isolated Python 3.12 environment using `torch 2.14.0.dev20260810+rocm7.2`: `torch.cuda.is_available() == True`, `device_count == 2`, and an actual tensor computation succeeded on each card. No `HSA_OVERRIDE_GFX_VERSION` was set or needed.

## What was installed and configured

Nothing was added system-wide; the account has Docker access but unattended `sudo` is unavailable.

- kind 0.32.0: `~/.local/bin/kind`
- kubectl 1.36.3: `~/.local/bin/kubectl`
- SkyPilot 0.13.0 with Kubernetes extras: `~/.local/share/skywright/skypilot-0.13.0`, exposed as `~/.local/bin/sky`
- Ubuntu's signed `socat` 1.8.0.0 package extracted without root under `~/.local/share/skywright/socat-1.8.0.0`, exposed as `~/.local/bin/socat`
- kind cluster `skywright-onprem`, Kubernetes 1.36.1, context `kind-skywright-onprem`
- The kind node receives `/dev/kfd` and `/dev/dri` through `~/.local/share/skywright/onprem/kind-config.yaml`.
- The plain AMD device plugin 1.31.0.10 is pinned by digest and applied from `~/.local/share/skywright/onprem/amd-device-plugin.yaml`. The AMD GPU Operator, cert-manager and driver lifecycle were not needed.
- The node is manually labelled `skypilot.co/accelerator=rx7800xt`.
- The kind node container uses Docker's `unless-stopped` restart policy. Docker itself is enabled at boot; reboot recovery is configured but not reboot-tested.

The earlier pod-egress failure did not recur with real Docker: Kubernetes and SkyPilot runtime setup both reached the internet successfully.

## What Kubernetes and SkyPilot report

Kubernetes publishes exactly:

```text
capacity:    amd.com/gpu: 2
allocatable: amd.com/gpu: 2
```

SkyPilot reports:

```text
Kubernetes GPUs
Context: kind-skywright-onprem
GPU       REQUESTABLE_QTY_PER_NODE  UTILIZATION
RX7800XT  1, 2                      2 of 2 free
```

The literal two-card resource specification is:

```yaml
resources:
  infra: k8s/kind-skywright-onprem
  accelerators: RX7800XT:2
  image_id: docker:rocm/pytorch:rocm7.2_ubuntu24.04_py3.12_pytorch_release_2.9.1@sha256:a3b65813621095e3389269417e963725b59310184588c9d2490d44e6e83fa01c
```

`sky launch` selected that exact context at `$0.00`, scheduled both GPUs, and completed a PyTorch job with `device_count=2`; computation succeeded on each `gfx1101` device. The test cluster was then torn down, leaving both cards free. Re-runnable manifests are in `~/.local/share/skywright/onprem/`.

## Negative findings and fragility

- `accelerators: 16GB+:1` does **not** match this node. SkyPilot expands the capability request through its static catalog (including `Radeon MI25` and `MI300X`, but not RX 7800 XT) and raises `ResourcesUnavailableError`. The literal `RX7800XT` name is mandatory.
- SkyPilot 0.13.0's credential check requires `socat` even with `kubernetes.networking: nodeport`; the private `socat` extraction is a workaround for that check.
- The device plugin logs harmless missing `current_compute_partition` / `current_memory_partition` warnings. Consumer RDNA cards do not expose the Instinct partition files, but both devices register healthy and workloads pass.
- This is a privileged kind node plus privileged device-plugin DaemonSet, not a conventional host-installed Kubernetes distribution. It avoids root changes and recurring cost, but is more fragile: recreating the kind cluster requires the saved config, reapplying the plugin, relabelling the node, and resetting the Docker restart policy.
- Reboot survival and behavior across host ROCm/kernel upgrades remain unverified.

## Credentials and state

- `~/.kube/config` (mode `0600`) contains the kind context's generated client key/certificate and CA data.
- `~/.sky/config.yaml` (mode `0600`) allows only `kind-skywright-onprem` and selects `nodeport` networking.
- SkyPilot's local API-server databases/logs live under `~/.sky/`; the CLI starts the local API server on demand.
- No cloud-provider, registry, or external-storage credentials were created.
